heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/MetricsCache.java

```diff
@@ -0,0 +1,187 @@
+    } else if ("LAST".equals(type)) {
+      return MetricsFilter.MetricAggregationType.LAST;
+    } else {
+      LOG.log(Level.SEVERE, "Unknown metrics type in metrics sinks " + type);
+      return MetricsFilter.MetricAggregationType.UNKNOWN;
+    }
+  }
+
+  /**
+   * sink publishes metrics and exceptions to this interface
+   *
+   * @param metrics message from sinks
+   */
+  public void AddMetric(TopologyMaster.PublishMetrics metrics) {
+    cache.AddMetricException(metrics);
+  }
+
+  /**
+   * for inside SLA process component query
+   *
+   * @param request query statement
+   * @return metric list
+   */
+  public MetricsCacheQueryUtils.MetricResponse GetMetrics(
+      MetricsCacheQueryUtils.MetricRequest request) {
+    return cache.GetMetrics(request, metricNameType);
+  }
+
+  /**
+   * for inside SLA process component query
+   *
+   * @param request query statement
+   * @return exception list
+   */
+  public MetricsCacheQueryUtils.ExceptionResponse GetExceptions(
+      MetricsCacheQueryUtils.ExceptionRequest request) {
+    return cache.GetExceptions(request);
+  }
+
+  /**
+   * compatible with tmaster interface
+   *
+   * @param request query request defined in protobuf
+   * @return query result defined in protobuf
+   */
+  public TopologyMaster.ExceptionLogResponse GetExceptions(
+      TopologyMaster.ExceptionLogRequest request) {
+    MetricsCacheQueryUtils.ExceptionRequest request1 = MetricsCacheQueryUtils.Convert(request);
+    MetricsCacheQueryUtils.ExceptionResponse response1 = cache.GetExceptions(request1);
+    TopologyMaster.ExceptionLogResponse response = MetricsCacheQueryUtils.Convert(response1);
+    return response;
+  }
+
+  private MetricsCacheQueryUtils.ExceptionResponse SummarizeException(
+      MetricsCacheQueryUtils.ExceptionResponse response1) {
+    return null;
+  }
+
+  /**
+   * compatible with tmaster interface
+   *
+   * @param request query statement defined in protobuf
+   * @return query result defined in protobuf
+   */
+  public TopologyMaster.ExceptionLogResponse GetExceptionsSummary(
+      TopologyMaster.ExceptionLogRequest request) {
+    MetricsCacheQueryUtils.ExceptionRequest request1 = MetricsCacheQueryUtils.Convert(request);
+    MetricsCacheQueryUtils.ExceptionResponse response1 = cache.GetExceptions(request1);
+    MetricsCacheQueryUtils.ExceptionResponse response2 = SummarizeException(response1);
+    TopologyMaster.ExceptionLogResponse response = MetricsCacheQueryUtils.Convert(response2);
+    return response;
+  }
+
+  /**
+   * compatible with tmaster interface
+   *
+   * @param request query statement defined in protobuf
+   * @return query result defined in protobuf
+   */
+  public TopologyMaster.MetricResponse GetMetrics(TopologyMaster.MetricRequest request) {
+    String componentName = request.getComponentName();
+    if (!cache.existComponentInstance(componentName, null)) {
+      TopologyMaster.MetricResponse.Builder builder =
```

Comment:
This builder logic is duplicated in this class. Better to put it in a private static helper that takes a message and returns a TopologyMaster.MetricResponse.